pkg/server/preprocessor/plugin.go

```diff
@@ -26,11 +26,29 @@ type EnsurePluginReadyPreprocessor struct {
 	PluginContext *plugin.Context
 }
 
-func (p *EnsurePluginReadyPreprocessor) Preprocess(payload *router.Payload, response *router.Response) int {
-	if !p.PluginContext.IsReady() {
-		log.Errorf("Request cannot be handled because plugins are unavailable at the moment.")
-		response.Err = skyerr.NewError(skyerr.PluginUnavailable, "plugins are unavailable at the moment")
-		return http.StatusServiceUnavailable
+func (p *EnsurePluginReadyPreprocessor) Preprocess(
+	payload *router.Payload,
+	response *router.Response,
+) int {
+	if p.PluginContext.IsReady() {
+		return http.StatusOK
 	}
-	return http.StatusOK
+
+	if p.PluginContext.IsInitialized() {
+		if !payload.HasMasterKey() || !payload.IsFromPlugin() {
+			response.Err = skyerr.NewError(
+				skyerr.PluginInitializing,
+				"Plugins are initializing at the moment",
+			)
+			return http.StatusServiceUnavailable
+		}
+
+		return http.StatusOK
+	}
+
+	response.Err = skyerr.NewError(
+		skyerr.PluginUnavailable,
+		"plugins are unavailable at the moment",
+	)
+	return http.StatusServiceUnavailable
```

Comment:
I think a test case for this preprocessor is needed as it before more complicated.